### PR TITLE
Engulfment tweaks and fixes

### LINKED ISSUE #3498: Health bar stuck in orange color after been engulfed, host killed, and killed because 30% damage.
image

### LINKED ISSUE #3499: Collection modified exception when changing patch
Godot stack trace:
```CSharp
Unhandled Exception:
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
  at System.Collections.Generic.List`1+Enumerator[T].MoveNextRare () [0x00013] in <1400ea11d6fd49d4939be686a9549fb4>:0
  at System.Collections.Generic.List`1+Enumerator[T].MoveNext () [0x0004a] in <1400ea11d6fd49d4939be686a9549fb4>:0
  at Microbe.ClearEngulfedObjects () [0x0006d] in D:\source\repos\Thrive\src\microbe_stage\Microbe.Contact.cs:524
  at MicrobeStage.UpdatePatchSettings (System.Boolean promptPatchNameChange) [0x0003d] in D:\source\repos\Thrive\src\microbe_stage\MicrobeStage.cs:614
  at MicrobeStage.SetupStage () [0x000ae] in D:\source\repos\Thrive\src\microbe_stage\MicrobeStage.cs:510
  at MicrobeStage._Ready () [0x0006b] in D:\source\repos\Thrive\src\microbe_stage\MicrobeStage.cs:104
  at (wrapper managed-to-native) Godot.NativeCalls.godot_icall_2_441(intptr,intptr,intptr,bool)
  at Godot.Node.AddChild (Godot.Node node, System.Boolean legibleUniqueName) [0x00001] in /root/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/Node.cs:527
  at SceneManager.SwitchToScene (Godot.Node newSceneRoot, System.Boolean keepOldRoot) [0x00031] in D:\source\repos\Thrive\src\engine\SceneManager.cs:49
  at EditorBase`2[TAction,TStage].OnEditorExitTransitionFinished () [0x00097] in D:\source\repos\Thrive\src\general\EditorBase.cs:763
  at TransitionManager+Sequence.StartNext () [0x000a1] in D:\source\repos\Thrive\src\gui_common\TransitionManager.cs:262
  at TransitionManager+Sequence.Process () [0x000ab] in D:\source\repos\Thrive\src\gui_common\TransitionManager.cs:232
  at TransitionManager._Process (System.Single delta) [0x00020] in D:\source\repos\Thrive\src\gui_common\TransitionManager.cs:76
```

[crash_save.zip](https://github.com/Revolutionary-Games/Thrive/files/9078525/quick_save_4.zip)

To reproduce:
1. Load this save
2. Change patch to the one on top of it
3. Confirm

There's an immediate ejection which modifies `engulfedObjects` at line 1820
```CSharp
Microbe.CompleteEjection() at D:\source\repos\Thrive\src\microbe_stage\Microbe.Contact.cs:line 1819
Microbe.EjectEngulfable() at D:\source\repos\Thrive\src\microbe_stage\Microbe.Contact.cs:line 1575
Microbe.ClearEngulfedObjects()
MicrobeStage.UpdatePatchSettings()
MicrobeStage.SetupStage()
MicrobeStage._Ready()
[Native Transition]
Node.AddChild()
SceneManager.SwitchToScene()
EditorBase<EditorAction, MicrobeStage>.OnEditorExitTransitionFinished()
TransitionManager.Sequence.StartNext()
TransitionManager.Sequence.Process()
TransitionManager._Process()
```

### PR DESCRIPTION
**Pull request by Kasterisk**
**Brief Description of What This PR Does**

Further engulfment tweaks and fixes. Follow up to #3106 

- Should fix game crash when changing patch while ingested something
- Make turning off engulf mode while still engulfing not spits out the object
- Non-digestible compounds are now absorbed but will be immediately vented out.
- Fix `CalculateNearbyWorldPosition()`
- Fix critical bug where `FloatingChunk.Compounds` is not saved
- Fix loading a save with ingested floating chunks making them disappear
- Fix health bar stuck in orange color
- Make player can't be engulfed while in godmode
- Make digestion not pause if a cell is on full capacity for a specific compound, instead the excess is ejected out to the environment
- The size of engulfed object scales with their digested amount, as a result the more digested an object is the less filled the ingested matter bar will be, allowing you to eat more things irrespective whether the digestion is finished or not.

**Related Issues**

Fixes #3499 
Fixes #3498 

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by 84634E1A607A** [comment]
Can confirm the first bug no longer appears. (used the same way to trigger the crash but failed)

**Comment by 84634E1A607A** [comment]
I'm stuck here, ammonia is full and I stopped digesting things I engulfed.
image

**Comment by 84634E1A607A** [comment]
Fine then.

**Comment by 84634E1A607A** [comment]
Things get worse now. 
image

**Comment by Kasterisk** [comment]
What's going on in that picture?

**Comment by 84634E1A607A** [comment]
You see, ammonia is full so I stopped digesting, but phosphate is 0 so I can't gain more by engulfing.

**Comment by Kasterisk** [comment]
Alright so the phosphate not increasing is because there aren't enough phosphates to absorb, not a bug. The other problem with the incorrect amount of ingested matter when loading a save seems to be that floating chunks are immediately in a fully digested state, this one is a bug.

**Comment by RacerBG** [comment]
Confirming that points 1, 2 and 3 are fixed.

**Comment by Kasterisk** [comment]
Thanks for testing.

> I have no idea about points 4 and 5 because I haven't seen them in my game.

They are just code specific details to the preceding points.

**Comment by RacerBG** [comment]
> Fix loading a save with ingested floating chunks making them disappear

Fixed.

**Comment by Kasterisk** [comment]
So to not frustrate players, I decided that digestion should not pause if cell is on full compounds capacity, instead the excess is ejected out to the environment. This is to avoid the player being stuck on digesting useless compounds.

Consider the following scenario: cell is on full engulfment (and compounds) capacity but none of the ingested objects contain vital compounds such as glucose or iron. What will happen is that the cell is basically stuck on absorbing useless compounds, it can't engulf other things until the useless compounds are depleted and digestion completes. The cell will probably die a slow death while it burns through its vital compounds reserve.

**Review by 84634E1A607A — APPROVED** [review]
Playtested. Good work!

**Review by RacerBG — APPROVED** [review]

File: src/general/StageHUDBase.cs [hunk 1]
```diff
@@ -903,33 +903,33 @@ protected Color GetCompoundDensityCategoryColor(float amount)
...
     ///   Updates the compound bars with the correct values.
     /// </summary>
-    protected virtual void UpdateCompoundBars()
+    protected virtual void UpdateCompoundBars(float delta)
     {
         var compounds = GetPlayerStorage();
 
         glucoseBar.MaxValue = compounds.GetCapacityForCompound(glucose);
-        glucoseBar.Value = compounds.GetCompoundAmount(glucose);
+        GUICommon.SmoothlyUpdateBar(glucoseBar, compounds.GetCompoundAmount(glucose), delta);
```
Comment (hhyyrylainen) [inline src/general/StageHUDBase.cs:911]: I really hope these don't cause similar bar stuck bugs as we've seen with the ATP and health bars...

File: src/microbe_stage/IEngulfable.cs [hunk 2]
```diff
@@ -9,6 +9,11 @@ public interface IEngulfable : IGraphicalEntity
     /// <summary>
     ///   The engulf size of this engulfable object.
     /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     This can also be made to scale with <see cref="DigestedAmount"/> for richer gameplay depth.
```
Comment (hhyyrylainen) [inline src/microbe_stage/IEngulfable.cs:14]: I'd reword this that some classes may scale this. Right now this sounds like some future thing but seems to be already implemented in FloatingChunk.

File: src/microbe_stage/MicrobeColony.cs [hunk 3]
```diff
@@ -69,14 +68,23 @@ public float HexCount
...
         {
-            if (hexCountDirty)
-                UpdateHexCount();
-            return usedIngestionCapacity;
+            var result = 0.0f;
+
+            foreach (var microbe in ColonyMembers)
+                result += microbe.UsedIngestionCapacity;
+
+            return result;
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeColony.cs:87]:
```suggestion
            return ColonyMembers.Sum(c => c.UsedIngestionCapacity);
```
One liner that should do the the same thing

**Comment by hhyyrylainen** [comment]
The changes pretty much make sense to me, though I made a few small comments. Seeing as the playtesting was positive, I think this can be merged once my comments are addressed, even though I didn't really have the time to look through the microbe class changes with as a finetooth comb as I would have liked.

File: src/microbe_stage/IEngulfable.cs (hunk 2 above)
Comment (Kasterisk) [reply src/microbe_stage/IEngulfable.cs:14]: Doesn't this have similar meaning?

File: src/microbe_stage/Microbe.Interior.cs [hunk 4]
```diff
@@ -1092,39 +1092,18 @@ private void SpawnEjectedCompound(Compound compound, float amount)
     /// </summary>
     private Vector3 CalculateNearbyWorldPosition()
     {
-        // The back of the microbe
-        var exit = Hex.AxialToCartesian(new Hex(0, 1));
-        var membraneCoords = Membrane.GetVectorTowardsNearestPointOfMembrane(exit.x, exit.z);
-
-        // Get the distance to eject the compounds
-        var ejectionDistance = Membrane.EncompassingCircleRadius;
+        var distance = Membrane.EncompassingCircleRadius;
```
[the remark this answers is not in the document]
Comment (Kasterisk) [reply src/microbe_stage/Microbe.Interior.cs:1095]: Probably yeah according to toxin emission code. Yes, will do.

File: src/general/StageHUDBase.cs (hunk 1 above)
Comment (Kasterisk) [reply src/general/StageHUDBase.cs:911]: This static method was made to fix that very problem.

File: src/microbe_stage/Microbe.Interior.cs [hunk 5]
```diff
@@ -1092,6 +1092,8 @@ private void SpawnEjectedCompound(Compound compound, float amount)
     /// </summary>
     private Vector3 CalculateNearbyWorldPosition()
     {
+        // OLD CODE
```
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.Interior.cs:1095]: This should explain why this is kept here. Also comparing the old logic to the new, the only bit that really is different is using Membrane.GetVectorTowardsNearestPointOfMembrane, besides that the old code could be rewritten to use your approach to the vector calculation, just my using a different distance.

Comment (Kasterisk) [reply src/microbe_stage/Microbe.Interior.cs:1095]: I decided to use `EncompassingCircleRadius` as the other I think might be a bit computationally expensive to be run each frame.

Comment (hhyyrylainen) [reply src/microbe_stage/Microbe.Interior.cs:1095]: That's related to the new code, right? Otherwise I don't understand what you are getting at.
Finding a membrane point for venting compounds is not that performance intensive, well at least I hope, because the externally positioned organelles *each* basically call that each frame due to the position problems they run into otherwise...

Comment (Kasterisk) [reply src/microbe_stage/Microbe.Interior.cs]: > That's related to the new code, right?

Yes. Well I think it's good to save a bit of performance with this.

Added little explanation as requested.
